cuda/factorization/par_ilu_kernels.cu

```diff
@@ -60,9 +65,92 @@ constexpr int default_block_size{512};
+    Array<bool> needs_change_device{exec, 1};
+    needs_change_device = needs_change_host;
+
+    auto cuda_old_values = as_cuda_type(mtx->get_const_values());
+    auto cuda_old_col_idxs = as_cuda_type(mtx->get_const_col_idxs());
+    auto cuda_old_row_ptrs = as_cuda_type(mtx->get_row_ptrs());
+    auto cuda_row_ptrs_add = as_cuda_type(row_ptrs_addition.get_data());
+
+    const dim3 block_dim{default_block_size, 1, 1};
+    const dim3 grid_dim{
+        static_cast<uint32>(ceildiv(
+            num_rows, static_cast<size_type>(block_dim.x / subwarp_size))),
+        1, 1};
+    if (is_sorted) {
+        kernel::find_missing_diagonal_elements<true, subwarp_size>
+            <<<grid_dim, block_dim>>>(
+                num_rows, num_cols, cuda_old_col_idxs, cuda_old_row_ptrs,
+                cuda_row_ptrs_add,
+                as_cuda_type(needs_change_device.get_data()));
+    } else {
+        kernel::find_missing_diagonal_elements<false, subwarp_size>
+            <<<grid_dim, block_dim>>>(
+                num_rows, num_cols, cuda_old_col_idxs, cuda_old_row_ptrs,
+                cuda_row_ptrs_add,
+                as_cuda_type(needs_change_device.get_data()));
+    }
+    needs_change_host = needs_change_device;
+    if (!needs_change_host.get_const_data()[0]) {
+        return;
+    }
+
+    prefix_sum(exec, cuda_row_ptrs_add, row_ptrs_size);
+    exec->synchronize();
+
+    IndexType total_additions{};
+    exec->get_master()->copy_from(
+        exec.get(), 1, cuda_row_ptrs_add + row_ptrs_size - 1, &total_additions);
+    size_type new_num_elems = static_cast<size_type>(total_additions) +
+                              mtx->get_num_stored_elements();
+
+
+    Array<ValueType> new_values{exec, new_num_elems};
+    Array<IndexType> new_col_idxs{exec, new_num_elems};
+    auto cuda_new_values = as_cuda_type(new_values.get_data());
+    auto cuda_new_col_idxs = as_cuda_type(new_col_idxs.get_data());
+
+    kernel::add_missing_diagonal_elements<subwarp_size>
+        <<<grid_dim, block_dim>>>(num_rows, cuda_old_values, cuda_old_col_idxs,
+                                  cuda_old_row_ptrs, cuda_new_values,
+                                  cuda_new_col_idxs, cuda_row_ptrs_add);
+
+    const dim3 grid_dim_row_ptrs_update{
+        static_cast<uint32>(
+            ceildiv(num_rows, static_cast<size_type>(block_dim.x))),
+        1, 1};
+    kernel::update_row_ptrs<<<grid_dim_row_ptrs_update, block_dim>>>(
+        num_rows + 1, cuda_old_row_ptrs, cuda_row_ptrs_add);
+
+    matrix::CsrBuilder<ValueType, IndexType> mtx_builder{mtx};
+    mtx_builder.get_value_array() = std::move(new_values);
+    mtx_builder.get_col_idx_array() = std::move(new_col_idxs);
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_PAR_ILU_ADD_DIAGONAL_ELEMENTS_KERNEL);
+
+
 template <typename ValueType, typename IndexType>
 void initialize_row_ptrs_l_u(
-    std::shared_ptr<const CudaExecutor> exec,
+    std::shared_ptr<const DefaultExecutor> exec,
```

Comment:
For consistency: Do we want to use this everywhere or revert these changes?

Reply:
I reverted the changes. You are right, updating just one file is inconsistent, we should update all or none.